### PR TITLE
Add minimal authentication to the API

### PR DESCRIPTION
**Pull request by simlmx**
_Best reviewed commit by commit_

The main purpose of this pull request is to add basic route protection. I only protect `/sites*` but we can trivially protect more routes as we wish. All that authentication logic is in the 3rd commit of this pull request.

To make my life easier, I did quite a big refactoring of the way we define the `app`. This is in the first commit. The big rationale behind the changes is to remove global variables. Python API frameworks have a tendency to use global variables everywhere (`app`, `db`, etc.) and it makes testing harder because we end up with a bunch of variables that are defined at `import`-time, and not at run-time, and so that are harder to change in tests. To summarize, instead of having:

```python
app = FastAPI()

app.get('/route')
def route():
    pass
```

now we have:

```python
def _add_routes(app):
    app.get('/route')
    def route():
        pass

def create_app():
    app = FastAPI()

    _add_routes(app)

    return app    
```

In the second commit, I organize all the environment variable into a single `Config` object. This way we can define this object once and then inject it where needed, like in the `create_app` function.

---------

You can view those two first commits as a Request For Comments. If y'all don't like them I can always revert them: they were less critical than I thought for the specific case of authentication, but I still think that they make the code clearer and better organized. Let me know what you think!

### CODE REVIEW

File: pv_site_api/app.py [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+        allow_credentials=True,
+        allow_methods=["*"],
+        allow_headers=["*"],
+    )
+
+
+def _create_site_router(*, is_fake: bool, get_session: Callable, auth: Callable) -> None:
+    router = APIRouter(prefix="/sites", dependencies=[Depends(auth)])
+
+    @router.get("/", response_model=PVSites)
```
Comment (simlmx) [inline pv_site_api/app.py:82]: The code is the same as before, but it is wrapped into function so that the routes are not defined at import-time, but only when we call the function. This way we can easily inject different dependencies when testing.

File: pv_site_api/main.py [hunk 2]
```diff
@@ -1,448 +1,23 @@
...
-"""
-
-origins = os.getenv("ORIGINS", "*").split(",")
-app.add_middleware(
-    CORSMiddleware,
-    allow_origins=origins,
-    allow_credentials=True,
-    allow_methods=["*"],
-    allow_headers=["*"],
+app = create_app(
```
Comment (simlmx) [inline pv_site_api/main.py:19]: Now the entry point is cleaner: we only define the dependencies and inject them into the `create_app` function.

File: pv_site_api/app.py (hunk 1 above)
Comment (peterdudfield) [reply pv_site_api/app.py:83]: hmm, I understand why you did this, but its not very neat, and I havent seen any other FastAPIs use this. 

[This](https://github.com/openclimatefix/nowcasting_api/blob/main/src/tests/test_gsp.py#L27) is how I ovverride the session in the other api

Could we do something similar for `auth`? Another way I got round it [was](https://github.com/openclimatefix/nowcasting_api/blob/main/src/auth_utils.py#L24) if `AUTH0_DOMAIN` and `AUTH0_API_AUDIENCE` not set return None - this seemed to get round auth when testing. 

I wonder if you can do something similar?

File: tests/test_generation.py [hunk 3]
```diff
@@ -8,15 +8,15 @@
...
+def test_pv_actual_fake(fake, client):
     response = client.get("/sites/fff-fff-fff/pv_actual")
     assert response.status_code == 200
 
     pv_actuals = MultiplePVActual(**response.json())
     assert len(pv_actuals.pv_actual_values) > 0
 
 
-def test_pv_actual_many_sites_fake(client, fake):
+def test_pv_actual_many_sites_fake(fake, client):
```
Comment (peterdudfield) [inline tests/test_generation.py:19]: why this change?

File: tests/test_auth.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        pass
+
+    yield TestClient(app)
+
+
+def _make_header(token):
+    return {"Authorization": f"Bearer {token}"}
+
+
+def test_auth_happy_path(trivial_client, config):
```
Comment (peterdudfield) [inline tests/test_auth.py:56]: I wonder if we should split the test into `unittest` and `intergration`. The unittest could have no aith on it, and the intergations could have the real auth on them

Comment (peterdudfield) [reply tests/test_auth.py:56]: Actually ive just seen you have monkey patched it, so maybe its fine

File: tests/conftest.py [hunk 5]
```diff
@@ -137,7 +144,12 @@ def statuses(db_session):
 
 @pytest.fixture()
 def fake(monkeypatch):
-    """Set up ENV VAR FAKE to 1"""
+    """Set up ENV VAR FAKE to 1
+
+    Note that this fixture must be included before the others because the
+    environment variables are read in the `config` fixture, on which many
+    other depend.
+    """
```
Comment (peterdudfield) [inline tests/conftest.py:152]: thanks

File: tests/test_generation.py (hunk 3 above)
Comment (peterdudfield) [reply tests/test_generation.py:19]: I read the comment above

File: pv_site_api/app.py [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+        environment=config.ENVIRONMENT,
+        traces_sampler=0.0 if config.ENVIRONMENT == "local" else traces_sampler,
+    )
+
+    _add_cors(app, origins=config.CORS_ORIGINS)
+
+    _add_base_routes(app, is_fake=config.FAKE, get_session=get_session)
+
+    site_router = _create_site_router(is_fake=config.FAKE, get_session=get_session, auth=auth)
+    app.include_router(site_router)
```
Comment (peterdudfield) [inline pv_site_api/app.py:437]: I can understand why you did this, 

I wonder if we should make sure the code is simple, and then have slightly more complex test setup, than, making the code more complicated. I think in general we should try and stick with the FastAPI way of doing things, as it will be easier to maintain, develop, e.t.c

File: pv_site_api/app.py (hunk 1 above)
Comment (simlmx) [reply pv_site_api/app.py:83]: We can use `dependency_override` too, it's just not needed if you inject it like I did.

I can put it back as it was before. This here was how I know to properly inject dependencies, in particular for tests. My main issue is with things like `load_dotenv()` flat in the file (which means that in your tests you'll use your non-test local `.env`, which can be unpredictable) and with all the global objects that then need to be monkeypatched. Injecting them is simpler!

It turns out that fastAPI has a [way to deal with configs][1], and it's to setup the equivalent of my `Config` object as a "Dendency" (using `Depends()`). It feels hackier to me but since it's their recommended approach we can go with that. It fixes the main issues while keeping the code closer to the fastAPI documentation.


[1]: https://fastapi.tiangolo.com/advanced/settings/#settings-in-a-dependency

**Comment by simlmx** [comment]
Reopened in #83 without the dependency injection fixes. Those can always be fixed later.
